src/ngraph/op/tensor_iterator.hpp

```diff
@@ -0,0 +1,110 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/node.hpp"
+#include "ngraph/op/parameter.hpp"
+
+namespace ngraph
+{
+    namespace op
+    {
+        struct InputSliceDescription
+        {
+            Output<Node> m_value;
+            std::ptrdiff_t m_start;
+            std::ptrdiff_t m_stride;
+            std::ptrdiff_t m_part_size;
+            std::ptrdiff_t m_end;
+            std::ptrdiff_t m_axis;
+        };
+
+        struct BodyConnection
+        {
+            Output<Node> m_initial_value;
+            Output<Node> m_body_value;
+        };
+
+        struct OutputConcatDescription
+        {
+            Output<Node> m_body_value;
+            std::ptrdiff_t m_start;
+            std::ptrdiff_t m_stride;
+            std::ptrdiff_t m_part_size;
+            std::ptrdiff_t m_end;
+            std::ptrdiff_t m_axis;
+        };
+
+        struct OutputSliceDescription
+        {
+            Output<Node> m_body_value;
+            int64_t m_index;
+        };
+
+        /// \brief  Iterate a body over tensors, accumulating into tensors
+        class TensorIterator : public util::FusedOp
+        {
+        public:
+            NGRAPH_API
+            static const std::string type_name;
+            const std::string& description() const override { return type_name; }
+        public:
```

Comment:
Why do you have two public sections here?

Reply:
Still a work in progress; not cleaned up yet.